### PR TITLE
MIRI and NIRcam PSF photometry with spacephot

### PR DESCRIPTION
**Pull request by orifox**
This notebook checklist has been made available to us by the [the Notebooks For All team](https://github.com/Iota-School/notebooks-for-all/blob/main/resources/event-hackathon/notebook-authoring-checklist.md).
Its purpose is to serve as a guide for both the notebook author and the technical reviewer highlighting critical aspects to consider when striving to develop an accessible and effective notebook.

### The First Cell

- [ ] The title of the notebook in a first-level heading (eg. `<h1>` or `# in markdown`).
- [ ] A brief description of the notebook.
- [ ] A table of contents in an [ordered list](https://www.markdownguide.org/basic-syntax/#ordered-lists) (`1., 2.,` etc. in Markdown).
- [ ] The author(s) and affiliation(s) (if relevant).
- [ ] The date first published.
- [ ] The date last edited (if relevant).
- [ ] A link to the notebook's source(s) (if relevant).

### The Rest of the Cells

- [ ] There is only one H1 (`#` in Markdown) used in the notebook.
- [ ] The notebook uses other [heading tags](https://www.markdownguide.org/basic-syntax/#headings) in order (meaning it does not skip numbers). 

## Text

- [ ] All link text is descriptive. It tells users where they will be taken if they open the link.
- [ ] All acronyms are defined at least the first time they are used. 
- [ ] Field-specific/specialized terms are used when needed, but not excessively.

## Code

- [ ] Code sections are introduced and explained before they appear in the notebook. This can be fulfilled with a heading in a prior Markdown cell, a sentence preceding it, or a code comment in the code section.
- [ ] Code has explanatory comments (if relevant). This is most important for long sections of code.
- [ ] If the author has control over the syntax highlighting theme in the notebook, that theme has enough color contrast to be legible.
- [ ] Code and code explanations focus on one task at a time. Unless comparison is the point of the notebook, only one method for completing the task is described at a time.

### Images

- [ ] All images (jpg, png, svgs) have an [image description](https://www.w3.org/WAI/tutorials/images/decision-tree/). This could be
    - [ ] [Alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt) (an `alt` property) 
    - [ ] [Empty alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt#decorative_images) for decorative images/images meant to be skipped (an `alt` attribute with no value)
    - [ ] Captions
    - [ ] If no other options will work, the image is decribed in surrounding paragraphs.

- [ ] Any [text present in images](https://www.w3.org/WAI/WCAG21/Understanding/images-of-text.html) exists in a text form outside of the image (this can be alt text, captions, or surrounding text.)

### Visualizations

- [ ] All visualizations have an image description. Review the previous section, Images, for more information on how to add it.
- [ ] [Visualization descriptions](http://diagramcenter.org/specific-guidelines-e.html) include
    - [ ] The type of visualization (like bar chart, scatter plot, etc.)
    - [ ] Title
    - [ ] Axis labels and range
    - [ ] Key or legend
    - [ ] An explanation of the visualization's significance to the notebook (like the trend, an outlier in the data, what the author learned from it, etc.)

- [ ] All visualizations and their parts have [enough color contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) ([color contrast checker](https://webaim.org/resources/contrastchecker/)) to be legible. Remember that transparent colors have lower contrast than their opaque versions.
- [ ] All visualizations [convey information with more visual cues than color coding](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html). Use text labels, patterns, or icons alongside color to achieve this.
- [ ] All visualizations have an additional way for notebook readers to access the information. Linking to the original data, including a table of the data in the same notebook, or sonifying the plot are all options.

### CODE REVIEW

File: notebooks/psf_photometry/NIRCam/nircam_spacephot.ipynb
Comment (mlibralato) [inline notebooks/psf_photometry/NIRCam/nircam_spacephot.ipynb:1]: <p>I have the exact same comments as for the MIRI notebook, so I will not add any below. I think that many parts are not clear, the usage of WebbPSF is uncertain and some of the results are puzzling.</p><p> </p><p>NIRCam images can be severely undersampled, so astrometry and photometry with them need a careful treatment. WebbPSF PSFs also do not include detector effects. Before making everything public, the undersampling systematics should be understood. The result of the PSF fit seems not that bad, but just looking at the subtracted image is not enough.</p><p> </p><p>I would choose a different set of images. The background in all images is significantly negative.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

**Comment by orifox** [comment]
> View / edit / reply to [this conversation](https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089638) on ReviewNB
> 
> _mlibralato commented on 2023-11-06T15:26:00Z_ ----------------------------------------------------------------
> 
> Please provide clear instructions about how to install the packages. If you install only space_phot, it retrieves the latest versions of all packages but new changes break some parts. The packages I had to change are the following:
> 
> * photutils==1.9.0 -> photutils==1.8.0
> * poppy==1.1.1 -> poppy==1.0.3
> * numexpr==2.8.7 -> numexpr==2.8.3

Fixed.

**Comment by orifox** [comment]
<p>Fixed.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089638'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089725'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089836'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089903'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p># Get PSF from WebbPSF</p><p># The function get_jwst_psf is a space_phot wrapper for the WebbPSF calc_psf function and uses a lot of the same keywords.</p><p># There are more advanced methods for generating your WebbPSF, but those are beyond the scope of this notebook.</p><p># The defaults used by get_jwst_psf in this notebook are:</p><p>&nbsp;&nbsp;# oversample=4</p><p>&nbsp;&nbsp;# normalize='last'</p><p>&nbsp;&nbsp;# Non-distorted PSF</p><p># Useful reference: <a href="https://webbpsf.readthedocs.io/en/latest/api/webbpsf.JWInstrument.html#webbpsf.JWInstrument.calc_psf" target="_blank">https://webbpsf.readthedocs.io/en/latest/api/webbpsf.JWInstrument.html#webbpsf.JWInstrument.calc_psf</a></p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089935'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>The goal is to calculate an upper limit on the blank part of the sky. Additional clarification has been written.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090114'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Need to ask Justin about the "Fix Background" Parameter setting here.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090180'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Fixed and hard-coded based on download script now included.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089672'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Good catch. Now automated and fixed. </p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089701'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>jwst datamodels now used.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089768'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Ok, updated. We are using Visit 006, which targets the A5V dwarf 2MASSJ17430448+6655015.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089870'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Ok, these are all great questions. A DEV note has been added. A DEV note is standard when more work needs to be done, but shouldn't hold up the publishing of the notebook. In this case there are a lot of things to do. First, space_phot needs updated documentation on what get_jwst_psf does. For now, I have added comments into the notebook referencing how it calls WebbPSF's calc_psf function using the defaults. That's sufficient for the purposes of this notebook. Generating more complex PSFs is beyond the scope of this notebook and should be built into the WebbPSF documentation or additional MIRI JDox documentation and/or notebooks. So that is Dev Note #2. </p><p> </p><p>For moving forward, we have added brief comments to temporarily address these issues with details about how the PSF is being generated.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089935'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Ok, a lot of comments to respond to here. Let me try my best.</p><p> </p><ol><li>Again, Dev notes are added for improved documentation on the metrics and how to set the parameters. Temporary notes have been added that address many of your questions.</li><li>The PSF oversubtraction could be caused by either of your options. I haven't had a chance for either Justin to optimize the fits or have someone help me come up with a better PSF model. To this end, I think finding the perfect PSF model is beyond the scope of this notebook. I agree its important to find a good model, but at this point, I haven't met anyone or read any documentation with the code to do it. So I have added a Dev note and commentary to this point. Nonetheless, the residual value is pretty good so the overall statistical error is relatively small.</li><li>The corner plots definitely need more explanation. That is captured in the Dev Note.</li><li>More description of the background fitting is needed. That is capture in the Dev Note.</li><li>The table shows both equatorial and pixel coordinates, but i think if you look at the table in the cell certain columns are not displayed. We think all values should be displayed to the user. No action has been taken.</li><li>The basic answer to the calibrated magnitude question is that the result is baked in. WebbPSF models are normalized (in this case) so that the fitted flux value is the same as the sum o fthe infinite PSF. So a fitted flux is not fitting the observed flux, its fitting TO the flux we see. This is important, but again, something that belongs in the documentation and not the notebook. We really need to beef up our PSF pages. This is captured in the Dev Note.</li><li>As for errors, the values areported are just statistical errors. For systematic errors one would need to plant/recover, but that is beyond the scope of this notebook. Nonetheless, we are clear that statistical errors are reported. This is no different than what Astropy's Photutils reports.</li></ol>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089990'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>At this point, we are taking several Level2 PSFs and resampling with the pipeline to make a Level2 PSF. We don't know how to turn off the output when that step is run.</p><p> </p><p>Regarding your questions, the Level2 PSFs are generated at the precise location of the source in each Level2 file to account for detector level effects. The resampling uses default resampling paramters. However, users should be aware that if they performed customized resampling for their Level2 data products, they should use similar resampling steps for their PSF below. We added a couple more comments and Dev Notes about these topics and believe that this comment is adequately addressed for the scope of this notebook.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090079'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Good catch. The averaging was all a bit old and not supposed to be in there for the exact reasons you give. The whole point to simultaneous fitting is to account for all of this. As a result, the flux and corresponding error take into account a single overall fit.&nbsp;Therefore, there is no need to average the resulting magnitudes or errors. They should all be the same to within their individual zero-point differences (typically &lt;1%).</p><p> </p><p>This is now captured in the documentation, notes, and Dev Notes. I think that addresses all of these concerns.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090021'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>These parameters were supposed to be optimized before we submitted the notebook, but that never got done. They are now optimized. In any case, PSF fitting does fit the background noise. That is the whole point to being able to calculate upper limits from PSF fitting. The values returned are the same as in Photutils. Photutils doesn't report a failed fit either. Instead, users need to be able to understand the goodness of fit from their error bars. That is now captured in the text in the notebook. </p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090145'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>done.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090322'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Thanks. Again we never got to finalizing the optimal paramters. That has now been done.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090180'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>I'm not exactly sure what you mean here, but this gives the opportunity of having the user define different background functions. I'm going to leave thi s and mark this resolved for now.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090212'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>space_phot current take in the skycoords positions, so we need this to run the package. I guess we could ask Justin to take the pixel coordinates in the future. I made that a DEV note for future work.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090252'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Good eye. </p><p> </p><p>Justin and I looked this over and believe there may be a bug in WebbPSF. We added a Dev note to open a github issue. But here are some of the details. In space_phot, the get_jwst_psf_from_grid calls webbpsf.gridded_library.CreatePSFLibrary(inst,inst.filter,&nbsp;num_psfs = num_psfs, detectors=st_obs.detector). There shouldn't be any keyword or parameter used here that gives the coordinates that you're seeing printed. If there is, we don't know how to do it. Maybe Marshall can help issue the right command. At the current time, we don't see any documentation on how to generate a more accurate PSF. We believe the PSFs are actually adequately sampled across the detector, but that WebbPSF is printing the wrong coordinates. Or, WebbPSF simply has a bug. But again, like I said, we opened a Dev Note and are going to investigate with the WebbPSF team. We don't want this to hold up the notebook.</p><p> </p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090280'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Ok, needed to optimize the parameters a bit. Should be better now. In any case, this loop shows the difficulty in doing PSF photometry on a wide variety of stars (brightness, distribution on the detector, etc) without visual inspection. This is true for all photometry programs. Especially when dealing with low SNR sources. Users should inspect the associated metrics and consider optimizing the parameters for specific stars of interest. Nonetheless, the success of the fits can always be quantified in with the associated error bars.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090388'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Ok, reference text added referring to earlier comments. Furthermore, details about empty plots are added.</p><p># It should be noted that empty plots correspond to LVL2 files with dither positions that do not cover that particular coordinate.</p><p> </p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090357'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: notebooks/psf_photometry/NIRCam/nircam_spacephot.ipynb
Comment (orifox) [reply notebooks/psf_photometry/NIRCam/nircam_spacephot.ipynb:1]: <p>Thanks! I'll return to these comments on NIRCam at a later date. If you have any good images to use, please let me know.</p>

**Comment by orifox** [comment]
<p>Thanks. This was an excellent review. Very thorough and you understood a lot of the intricacies. I've tried to address all of your comments, and I think I have to within the scope of this notebook. Particularly how to use the results to determine the goodness of fit. Remaining are a number of Developer Notes (Action Items) to significantly improve the documentation for WebbPSF, Photutils, JDox, and of course, Space_Phot. We'll need to add a lot of that, but I think that is a much larger effort beyond the scope of this notebook. At the current time, users have no examples on how to do PSF photometry. Particularly no examples on how to generate even the most basic JWST PSFs and use them to obtain PSF photometry. So I think if we can confirm the photometry results here are indeed correct and have reliable error bars, we can move forward. Particulary given the PSF notebooks that are currently available. What would be good is to understand within the scope of all of these limitations, what you think the best way to move forward is. </p><p> </p><p>Please remember, the only PSF photometry that users are currently being directed to is by Matteo from 3 years ago:</p><p><a href="https://spacetelescope.github.io/jdat_notebooks/notebooks/psf_photometry/NIRCam_PSF_Photometry_Example.html" target="_blank">https://spacetelescope.github.io/jdat_notebooks/notebooks/psf_photometry/NIRCam_PSF_Photometry_Example.html</a></p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795089604'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>I'm not sure how else to say it. The goal is to calculate an upper limit. Let's say you want to try to determine how bright a source is, but you can't see it in a particular filter. Then you need to calculate an upper limit. How do you normally calculate upper limits?</p><p> </p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090114'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Did you pull the latest PR? I'm not sure if the images generated above are what you're still seeing? Or if that's old? I don't see that in my current notebook.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090388'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by mlibralato** [comment]
<p>I think you mean lower limit. We add many sources of various brightness across the entire FoV and try to measure them. At a certain threshold, you recover less stars, let's say 50%. Then you go X mag fainter and you recover only 20% and so on. These values tell you what you can measure in an image.</p><p> </p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090114'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by orifox** [comment]
<p>Right, that's how you would traditionally do it over a field. But in many cases, and supernovae in particular, you have a very complicated background in a very specific region of the detector. Dropping a source at that position is certainly one way to do it. But doing PSF photometry with the "Upper Limit" mode turned on essentially does this for you. And with less systematic error. Also, in general, we call this an "Upper Limit". Because the source could be fainter, but not brighter. If it were brighter, we would have detected it.</p>

---

 <a href='https://app.reviewnb.com/spacetelescope/jdat_notebooks/pull/173/discussion/#comment-1795090114'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by camipacifici** [comment]
@orifox are these ready to go through technical review?

**Comment by orifox** [comment]
Ok, believe it or not, I finally did a review. The notebook results match pretty well, although admittedly not when they have to be run on batches of stars. The fine tuning required for space_phot is not efficient compared to the generic PSF photometry in photutils. This should probably be accounted for. One more iteration with Justin Pierel (author of space_phot) and @larrybradley wouldn't hurt (to determine which photometry is actually correct). But I can't commit to a time frame. So, at the moment, I rather explain it in the documentation, which I did. Ultimately, we may want a full page comparing/contrasting results. Here are my raw numbers.

Single Star

space_phot (lvl2) - 17.27 +/ - 0.0006
space_phot (lvl3) - 17.34 +/- 0.0007
Photutils - 17.26 +/- 0.002

Upper limit

space_phot (lvl2) - 25.68
space_phot (lvl3) - 25.92
Photutils - 25.77

Large Sample (In space_phot notebook, use following mask from photutils)
array([False,  True,  True,  True,  True, False,  True,  True,  True, True,  True,  True,  True])

space_phot (lvl2)

19.41 (0.006), 18.93 (0.005), 19.11 (0.006), 19.87 (0.010), 18.98 (0.005), 18.86 (0.005), 17.86 (0.003), 
17.97 (0.003), 18.55 (0.004), 19.30 (0.006), 17.99 (0.003)

space_phot (lvl3)

19.53 (0.003), 19.59 (0.003), 19.66 (0.003), 19.92 (0.006), 19.47 (0.003), 19.43 (0.003), 18.31 (0.002), 18.20 (0.002), 18.76 (0.002), 19.55 (0.003), 18.09 (0.002)

Photutils

19.42 (0.006), 18.71 (0.005), 18.97 (0.005), 19.11 (0.005), 18.46 (0.004), 19.03 (0.005), 18.92 (0.005), 17.90 (0.003), 18.00 (0.003), 18.62 (0.004), 17.99 (0.003)

**Comment by camipacifici** [comment]
Thank you very much @orifox!
@larrybradley, @mlibralato, are you satisfied with the updates and changes?
If you approve, we can send this to technical review.

**Comment by larrybradley** [comment]
Thanks, @orifox !
@camipacifici  - The `photutils` notebook looks good to me!

I looked into the `photutils` and `space_phot` comparison for the batches-of-stars example.  The differences that you quote above are actually because they are comparing different sources (the ordering is different).  I matched the sources based on their sky coordinates, and the magnitude agreement is excellent (see table below).  The comparison here is for the L2 data.  Another difference is that the `photutils` measurements are on only 1 L2 image, while I think the `space_phot` ones are the average of several L2 images.  Even without averaging in `photutils`, the agreement is great.

```
id          photutils_skcoord            spacephot_skycoord  skycoord_separation    photutils_mag    spacephot_mag       delta_mag      
                  deg,deg                      deg,deg               arcsec             mag(AB)          mag(AB)             mag         
--- ------------------------------------ -------------------- -------------------- ------------------ ------------- ---------------------
  1 80.70519471631073,-69.43753052941372 80.705193,-69.437533 0.009155037805485515   19.4215510154654       19.4137   -0.0078510154654019
  2 80.70480422819078,-69.44154405866392 80.704803,-69.441539 0.018277259194666006  19.11212668661077     19.107974  -0.00415268661077306
  3 80.70684588574794,-69.44434384295003 80.706852,-69.444343  0.00830297210447494 18.459203882785665      18.43691 -0.022293882785664465
  4  80.7213479879159,-69.44277479471901 80.721344,-69.442775 0.005095059299705519 19.028584737840696     18.982574  -0.04601073784069598
  5 80.72822954229039,-69.44592321158235  80.72824,-69.445923 0.013239726949403684 18.916197948146284     18.864805 -0.051392948146283146
  6 80.73141545916762,-69.44867247513723  80.73142,-69.448675 0.010749423666650859 17.898508125170252      17.85799  -0.04051812517025155
  7 80.72802026197112,-69.45134076578792  80.72802,-69.451343 0.008049972599883128 17.994802238488514     17.965346 -0.029456238488513975
  8 80.73747527736528,-69.44890734007218  80.737478,-69.44891 0.010175140025633956 18.619265512592772     18.553407   -0.0658585125927722
  9 80.74188340028199,-69.44834570133264 80.741885,-69.448348 0.008518582763235714 17.994906648279926     17.996318 0.0014113517200726733
```
One 9 of the 13 source matched.  2 are missing because they are off the first L2 image.  2 other sources did not match.

**Comment by camipacifici** [comment]
Lovely! Can we send this to technical review?

**Comment by larrybradley** [comment]
@camipacifici The `photutils` notebook is ready for technical review.  I didn't review the `space_phot` notebook.

**Comment by mlibralato** [comment]
@camipacifici I looked at the notebooks, without running them, and the documentation has significantly improved. I think you can send the `space_phot` notebook is ready for the technical review.

**Comment by camipacifici** [comment]
Thank you @mlibralato!!

**Comment by bhilbert4** [comment]
I'm meeting with @orifox next week to talk about photometry notebooks in general, as well as this particular notebook. It sounds like there may have been some recent updates to space_phot that may result in updates to this notebook.

**Review by gibsongreen — APPROVED** [review]
All execution, validation, and styling checks pass, ready to merge after rebasing and updating to new repository structure

**Pull request merged**
